### PR TITLE
Update docs about Sequence mutability

### PR DESCRIPTION
**Pull request by anderspitman**
Underlying sequences are immutable, but metadata and positional
metadata are mutable

### CODE REVIEW

**Comment by anderspitman** [comment]
Still working on IO stuff, but wanted to make sure this is looking ok.

**Comment by jairideout** [comment]
This looks great so far!

File: skbio/io/format/fasta.py [hunk 1]
```diff
@@ -84,9 +84,10 @@
 ~~~~~~~~~~~~~~~
 Each sequence header consists of a single line beginning with a greater-than
 (``>``) symbol. Immediately following this is a sequence identifier (ID) and
-description separated by one or more whitespace characters. Both sequence ID
-and description are optional and are represented as the empty string (``''``)
-in scikit-bio's objects if they are not present in the header.
+description separated by one or more whitespace characters. The sequence ID and
+description are stored in the sequence metadata attribute. Both are optional.
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: can you mark `metadata` in backticks? since it's an attribute you're referencing

Comment (jairideout) [inline skbio/io/format/fasta.py]: and note what the keys are (`'id'`, `'description'`)

File: skbio/io/format/fasta.py [hunk 2]
```diff
@@ -126,23 +127,21 @@
 lines. The sequence data (i.e., nucleotides or amino acids) are stored using
 the standard IUPAC lexicon (single-letter codes).
 
-.. note:: scikit-bio supports both upper and lower case characters. Both ``-``
+.. note:: scikit-bio supports both upper and lower case characters,
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: , -> .

File: skbio/io/format/fasta.py [hunk 3]
```diff
@@ -126,23 +127,21 @@
...
+   For reading, this functionality depends on the type of object the data is
+   being read in to. For ``Sequence``
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: in to -> into

File: skbio/io/format/fasta.py [hunk 4]
```diff
@@ -126,23 +127,21 @@
...
+   objects, sciki-bio doesn't care about the case. However, for other object
+   types, such as ``DNA``, ``RNA``, and ``Protein``, the `lowercase` parameter
```
Comment (ebolyen) [inline skbio/io/format/fasta.py]: link out with 
:class:\`skbio.DNA\`, etc

File: skbio/io/format/fasta.py [hunk 5]
```diff
@@ -126,23 +127,21 @@
 lines. The sequence data (i.e., nucleotides or amino acids) are stored using
 the standard IUPAC lexicon (single-letter codes).
 
-.. note:: scikit-bio supports both upper and lower case characters. Both ``-``
+.. note:: scikit-bio supports both upper and lower case characters,
+   For reading, this functionality depends on the type of object the data is
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: remove "for reading" since you're not talking about writing herej

File: skbio/io/format/fasta.py [hunk 6]
```diff
@@ -126,23 +127,21 @@
...
+   being read in to. For ``Sequence``
+   objects, sciki-bio doesn't care about the case. However, for other object
+   types, such as ``DNA``, ``RNA``, and ``Protein``, the `lowercase` parameter
+   must be used to control case functionality. Refer to the documentation for
+   the constructors for details. Both ``-``
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: move gap char discussion to its own paragraph

File: skbio/io/format/fasta.py (hunk 1 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Both done

File: skbio/io/format/fasta.py [hunk 7]
```diff
@@ -248,6 +250,10 @@
   behavior is to not split sequence data or quality scores across multiple
   lines.
 
+- ``lowercase``: Boolean array of the same length as the sequence, which
```
Comment (jairideout) [inline skbio/io/format/fasta.py]: document the string behavior

File: skbio/io/format/fasta.py (hunk 3 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: I did it that way in several places. Are you sure into is correct? Into indicates movement, we're talking about actually placing it inside the object, which deals with location

File: skbio/io/format/fastq.py [hunk 8]
```diff
@@ -83,14 +85,16 @@
    scikit-bio will ignore leading and trailing whitespace characters on each
    line while reading.
 
-   Whitespace characters are not allowed within sequence data or quality
-   scores.
+.. note:: Validation may be performed depending on the type of object the data
+   is being read in to. This behavior matches that of FASTA files.
```
Comment (jairideout) [inline skbio/io/format/fastq.py]: in to -> into

File: skbio/io/format/fasta.py (hunk 2 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Fixed.

File: skbio/io/format/fastq.py (hunk 8 above)
Comment (jairideout) [inline skbio/io/format/fastq.py]: FASTA files -> the FASTA reader

**Comment by jairideout** [comment]
@ebolyen and I are done reviewing, looks good just a few minor comments

File: skbio/io/format/fasta.py (hunk 4 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: It formats them differently but doesn't appear to make links

File: skbio/io/format/fasta.py (hunk 5 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Done

File: skbio/io/format/fasta.py (hunk 6 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Done

File: skbio/io/format/qseq.py [hunk 9]
```diff
@@ -47,19 +47,27 @@
 
 For more details please refer to the CASAVA documentation [1]_.
 
+.. note:: When a QSeq file is read in to a scikit-bio object, the object's
```
Comment (jairideout) [inline skbio/io/format/qseq.py]: in to -> into

File: skbio/io/format/qseq.py [hunk 10]
```diff
@@ -47,19 +47,27 @@
...
+   `metadata` attribute is automatically populated with data corresponding
+   to the names above.
+
+.. note:: `lowercase` functionality is supported when reading QSeq files,
+   depending on the object type the file is being read in to. Refer to
```
Comment (jairideout) [inline skbio/io/format/qseq.py]: in to -> into

File: skbio/io/format/fasta.py (hunk 7 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Done

File: skbio/io/format/fastq.py (hunk 8 above)
Comment (anderspitman) [inline skbio/io/format/fastq.py]: Fixed

File: skbio/io/format/fasta.py (hunk 3 above)
Comment (anderspitman) [inline skbio/io/format/fasta.py]: Fixed

File: skbio/io/format/fasta.py (hunk 4 above)
Comment (jairideout) [inline skbio/io/format/fasta.py]: try :class:\`skbio.sequence.DNA\`

File: skbio/sequence/_rna.py [hunk 11]
```diff
@@ -39,9 +39,15 @@ class RNA(IUPACSequence, NucleotideMixin):
         work or behave as expected.** Only turn off validation if you are
         certain that the sequence characters are valid. To store sequence data
         that is not IUPAC-compliant, use ``Sequence``.
-    case_insenstive : bool, optional
+    lowercase : bool or str, optional
         If ``True``, lowercase sequence characters will be converted to
-        uppercase characters in order to be valid IUPAC RNA characters.
+        uppercase characters in order to be valid IUPAC DNA characters. If
```
Comment (jairideout) [inline skbio/sequence/_rna.py]: DNA -> RNA

File: skbio/sequence/_protein.py [hunk 12]
```diff
@@ -39,9 +39,15 @@ class Protein(IUPACSequence):
         will work or behave as expected.** Only turn off validation if you are
         certain that the sequence characters are valid. To store sequence data
         that is not IUPAC-compliant, use ``Sequence``.
-    case_insenstive : bool, optional
+    lowercase : bool or str, optional
         If ``True``, lowercase sequence characters will be converted to
-        uppercase characters in order to be valid IUPAC protein characters.
+        uppercase characters in order to be valid IUPAC DNA characters. If
```
Comment (jairideout) [inline skbio/sequence/_protein.py]: DNA -> protein

File: skbio/sequence/_rna.py (hunk 11 above)
Comment (anderspitman) [inline skbio/sequence/_rna.py]: Fixed

File: skbio/sequence/_protein.py (hunk 12 above)
Comment (anderspitman) [inline skbio/sequence/_protein.py]: Fixed

**Comment by jairideout** [comment]
ready for merge when tests pass :shipit:

**Comment by jairideout** [comment]
:evanthomas:

**Comment by ebolyen** [comment]
:8ball: :gun: :game_die:

**Pull request merged**